File: m4/Act_IDL/src/py_add_three_ints/py_add_three_ints/add_three_ints_server.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from interfaces.srv import AddThreeInts


class AddThreeIntsServer(Node):
    def __init__(self):
        super().__init__('add_three_ints_server')

        self.service = self.create_service(
            AddThreeInts,
            'add_three_ints',
            self.handle_add_three_ints
        )

        self.get_logger().info('AddThreeInts server is ready.')

    def handle_add_three_ints(self, request, response):
        response.sum = request.a + request.b + request.c

        self.get_logger().info(
            f'Request received: a={request.a}, b={request.b}, c={request.c}'
        )
        self.get_logger().info(
            f'Response sent: sum={response.sum}'
        )

        return response


def main(args=None):
    rclpy.init(args=args)
    node = AddThreeIntsServer()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: m4/Act_IDL/src/cpp_add_three_ints/src/add_three_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "interfaces/srv/add_three_ints.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class AddThreeIntsServer : public rclcpp::Node
{
public:
  AddThreeIntsServer()
  : Node("cpp_add_three_ints_server")
  {
    service_ = this->create_service<interfaces::srv::AddThreeInts>(
      "add_three_ints",
      std::bind(&AddThreeIntsServer::handle_service, this, _1, _2)
    );

    RCLCPP_INFO(this->get_logger(), "C++ AddThreeInts server is ready.");
  }

private:
  void handle_service(
    const std::shared_ptr<interfaces::srv::AddThreeInts::Request> request,
    std::shared_ptr<interfaces::srv::AddThreeInts::Response> response)
  {
    response->sum = request->a + request->b + request->c;

    RCLCPP_INFO(
      this->get_logger(),
      "Request received: a=%ld, b=%ld, c=%ld",
      request->a, request->b, request->c
    );

    RCLCPP_INFO(
      this->get_logger(),
      "Response sent: sum=%ld",
      response->sum
    );
  }

  rclcpp::Service<interfaces::srv::AddThreeInts>::SharedPtr service_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AddThreeIntsServer>());
  rclcpp::shutdown();
  return 0;
}
